File: 2022/cpp/day8.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

int main() {
    ifstream infile;
    infile.open("input/day8.txt", ios::in);
    string cur;
    vector<string> trees;
    int count = 0;
    while (getline(infile, cur)) trees.push_back(cur);
    int row = trees.size(), col = trees[0].size(), pre;
    //part1
    bool see[row][col];
    for (int i = 0; i < row; i++) {
        for (int j = 0; j < col; j++)  see[i][j] = false;
    }
    for (int i = 0; i < row; i++) {
        pre = 0;
        for (int j = 0; j < col; j++) {
            if (pre == 9) break;
            if (pre < trees[i][j]-'0') {
                see[i][j] = true;
                pre = trees[i][j]-'0';
            }
        }
    }
    for (int i = 0; i < row; i++) {
        pre = 0;
        for (int j = col-1; j > 0; j--) {
            if (pre == 9) break;
            if (pre < trees[i][j]-'0'){
                see[i][j] = true;
                pre = trees[i][j]-'0';
            }
        }
    }
    for (int j = 0; j < col; j++) {
        pre = 0;
        for (int i = 0; i < row; i++) {
            if (pre == 9) break;
            if (pre < trees[i][j]-'0') {
                see[i][j] = true;
                pre = trees[i][j]-'0';
            }
        }
    }
    for (int j = 0; j < col; j++) {
        pre = 0;
        for (int i = row-1; i > 0; i--) {
            if (pre == 9) break;
            if (pre < trees[i][j]-'0') {
                see[i][j] = true;
                pre = trees[i][j]-'0';
            }
        }
    }
    for (int i = 0; i < row; i++) {
        for (int j = 0; j < col; j++) {
            if (i == 0 || i == row-1 || j == 0 || j == col-1) see[i][j] = true;
            if (see[i][j]) count++;
        }
    }
    cout << count << '\n';
    //part2
    int mx = 0;
    for (int i = 1; i < row-1; i++) {
        for (int j = 1; j < col-1; j++) {
            int up = 1, down = 1, left = 1, right = 1;
            while (i-up > 0 && trees[i-up][j] < trees[i][j]) up++;
            while (i+down < row-1 && trees[i+down][j] < trees[i][j]) down++;
            while (j-left > 0 && trees[i][j-left] < trees[i][j]) left++;
            while (j+right < col-1 && trees[i][j+right] < trees[i][j]) right++;
            mx = max(mx, up * down * left * right);
        }
    }
    cout << mx << '\n';
    return 0;
}

File: 2022/py/day8.py
with open("/Users/yetingzhi/Documents/GIT/adventofcode/input/day8.txt", "r") as f:
    trees = [list(map(int, line)) for line in f.read().splitlines()]
    count = 0
    mx = 0
    for i in range(len(trees)):
        for j in range(len(trees[i])):
            see = 0
            score = 1
            flag = 0
            for x in range(i-1, -1, -1):
                if trees[x][j] >= trees[i][j]:
                    score *= i-x
                    flag = 1
                    break
            if flag == 0: 
                score *= i
                see = 1
            flag = 0
            for x in range(i+1, len(trees)):
                if trees[x][j] >= trees[i][j]:
                    score *= x-i
                    flag = 1
                    break
            if flag == 0: 
                score *= len(trees)-i-1
                see = 1
            flag = 0
            for y in range(j-1, -1, -1):
                if trees[i][y] >= trees[i][j]:
                    score *= j-y
                    flag = 1
                    break
            if flag == 0: 
                score *= j
                see = 1
            flag = 0
            for y in range(j+1, len(trees[i])):
                if (trees[i][y] >= trees[i][j]):
                    score *= y-j
                    flag = 1
                    break
            if flag == 0: 
                score *= len(trees[i])-j-1
                see = 1
            count += see
            if score > mx: mx = score
                    
    print(count)
    print(mx)